auth/common.hh

```diff
@@ -95,12 +96,61 @@ future<> announce_mutations_with_batching(
         std::function<mutations_generator(api::timestamp_type& t)> gen,
         seastar::abort_source* as);
 
-// Execute update query via group0 mechanism, mutations will be applied on all nodes.
+// FIXME(mmal): remove
 future<> announce_mutations(
         cql3::query_processor& qp,
         ::service::raft_group0_client& group0_client,
         const sstring query_string,
         std::vector<data_value_or_unset> values,
         seastar::abort_source* as);
 
+// mutations_collector is used to gather mutations which are side effects
+// of functions execution. They need to be announced under single guard
+// for atomicity. As functions which produce mutations may embed each other
+// we need to decouple announcing step to a common external place here.
+// It also supports generator callbacks to avoid holding too many mutations
+// in memory as number of mutations may depend on user's data.
+// There is no guarantee about ordering of mutations.
+// If size of mutations is too big for raft too handle under single guard
+// atomicity guarantee may be violated.
+class mutations_collector {
+public:
+    using generator_func = std::function<mutations_generator(api::timestamp_type& t)>;
+private:
+    std::vector<mutation> _muts;
+    std::vector<generator_func> _generators;
+    std::optional<::service::group0_guard> _guard;
+public:
+    mutations_collector(::service::group0_guard g) : _guard(std::move(g)) {}
+    // This constructor is used to handle both legacy and current paths
+    // as there is no guard for legacy and the whole class is only passed
+    // through to simplify the code.
+    mutations_collector(std::optional<::service::group0_guard> g) {
+        if (g) {
+            _guard = std::move(*g);
+        }
+    };
+    ~mutations_collector();
+
+    mutations_collector(mutations_collector&) = delete;
+    mutations_collector(mutations_collector&&) = default;
+
+    // gets timestamp which should be used when building mutations
+    api::timestamp_type write_timestamp() const;
+
+    void add_mutation(mutation m);
+    void add_mutations(std::vector<mutation> ms);
+    void add_generator(generator_func f);
+    // announce commits all data, must be called only once. Nop if there was no guard provided.
```

Comment:
Outdated comment? The guard is implicit and it is an internal error if it is missing.

Reply:
yes